test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/VisualStudioCredentialProviderTests.cs

```diff
@@ -0,0 +1,128 @@
+using NuGet.Configuration;
+using NuGet.Credentials;
+using NuGet.VisualStudio;
+using Test.Utility.Threading;
+using Xunit;
+
+namespace NuGet.PackageManagement.VisualStudio.Test
+{
+    [Collection(DispatcherThreadCollection.CollectionName)]
+    public class VisualStudioCredentialProviderTests
+    {
+        private static readonly Uri _uri = new Uri("http://unit.test");
+
+        private readonly DispatcherThreadFixture _fixture;
+
+        public VisualStudioCredentialProviderTests(DispatcherThreadFixture fixture)
+        {
+            _fixture = fixture;
+        }
+
+        [Fact]
+        public void Constructor_ThrowsForNullWebProxyService()
+        {
+            var exception = Assert.Throws<ArgumentNullException>(() => new VisualStudioCredentialProvider(null));
+
+            Assert.Equal("webProxyService", exception.ParamName);
+        }
+
+        [Fact]
+        public void Id_IsRandomPerInstance()
+        {
+            const int trials = 3;
+            var hashset = new HashSet<string>();
+
+            for (var i = 0; i < trials; ++i)
+            {
+                hashset.Add(new VisualStudioCredentialProvider(Mock.Of<IVsWebProxy>()).Id);
+            }
+
+            Assert.Equal(trials, hashset.Count);
+        }
+
+        [Fact]
+        public async Task GetAsync_ThrowsForNullUri()
+        {
+            var provider = new VisualStudioCredentialProvider(Mock.Of<IVsWebProxy>());
+
+            var exception = await Assert.ThrowsAsync<ArgumentNullException>(
+                () => provider.GetAsync(
+                    uri: null,
+                    proxy: Mock.Of<IWebProxy>(),
+                    type: CredentialRequestType.Unauthorized,
+                    message: "a",
+                    isRetry: false,
+                    nonInteractive: true,
+                    cancellationToken: CancellationToken.None));
+
+            Assert.Equal("uri", exception.ParamName);
+        }
+
+        [Fact]
+        public async Task GetAsync_ThrowsIfCancelled()
+        {
+            var provider = new VisualStudioCredentialProvider(Mock.Of<IVsWebProxy>());
+
+            await Assert.ThrowsAsync<OperationCanceledException>(
+                () => provider.GetAsync(
+                    _uri,
+                    proxy: null,
+                    type: CredentialRequestType.Unauthorized,
+                    message: "a",
+                    isRetry: false,
+                    nonInteractive: true,
+                    cancellationToken: new CancellationToken(canceled: true)));
+        }
+
+        [Fact]
+        public async Task GetAsync_CallsWebProxyWithDefaultCredentialsState()
+        {
+            var vsWebProxy = new Mock<IVsWebProxy>(MockBehavior.Strict);
+            var newState = (uint)__VsWebProxyState.VsWebProxyState_NoCredentials;
+            var expectedCredentials = new NetworkCredential();
+
+            NuGetUIThreadHelper.SetCustomJoinableTaskFactory(_fixture.JoinableTaskFactory);
```

Comment:
Due to parallel test execution this statement may cause race conditions. I'd recommend initialize the provider with an instance of JTF passed down in a constructor.